certora/specs/ERC20Wrapper.spec

```
@@ -0,0 +1,190 @@
+*/
+invariant totalSupplyIsSmallerThanUnderlyingBalance()
+    totalSupply() <= underlyingBalanceOf(currentContract)  &&
+    underlyingBalanceOf(currentContract) <= underlyingTotalSupply() &&
+    underlyingTotalSupply() <= max_uint256
+    {
+        preserved {
+            requireInvariant totalSupplyIsSumOfBalances;
+            require underlyingBalanceOf(currentContract) <= underlyingTotalSupply();
+        }
+        preserved depositFor(address account, uint256 amount) with (env e) {
+            require e.msg.sender != currentContract;
+            sumOfUnderlyingBalancesLowerThanUnderlyingSupply(e.msg.sender, currentContract);
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: depositFor liveness and effects                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule depositFor(env e) {
+    require nonpayable(e);
+
+    address sender = e.msg.sender;
+    address receiver;
+    address other;
+    uint256 amount;
+
+    // sanity
+    require currentContract != sender;
+    require currentContract != underlying();
+    requireInvariant totalSupplyIsSumOfBalances;
+    requireInvariant totalSupplyIsSmallerThanUnderlyingBalance;
+    sumOfUnderlyingBalancesLowerThanUnderlyingSupply(currentContract, sender);
+
+    uint256 balanceBefore                   = balanceOf(receiver);
+    uint256 supplyBefore                    = totalSupply();
+    uint256 senderUnderlyingBalanceBefore   = underlyingBalanceOf(sender);
+    uint256 senderUnderlyingAllowanceBefore = underlyingAllowanceToThis(sender);
+    uint256 wrapperUnderlyingBalanceBefore  = underlyingBalanceOf(currentContract);
+    uint256 underlyingSupplyBefore          = underlyingTotalSupply();
+
+    uint256 otherBalanceBefore              = balanceOf(other);
+    uint256 otherUnderlyingBalanceBefore    = underlyingBalanceOf(other);
+
+    depositFor@withrevert(e, receiver, amount);
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> (
+        sender   != 0                             && // invalid sender
+        receiver != 0                             && // invalid receiver
+        amount   <= senderUnderlyingBalanceBefore && // deposit doesn't exceed balance
+        amount   <= senderUnderlyingAllowanceBefore  // deposit doesn't exceed allowance
+    );
+
+    // effects
+    assert success => balanceOf(receiver)                  == balanceBefore                  + amount;
+    assert success => totalSupply()                        == supplyBefore                   + amount;
+    assert success => underlyingBalanceOf(currentContract) == wrapperUnderlyingBalanceBefore + amount;
+    assert success => underlyingBalanceOf(sender)          == senderUnderlyingBalanceBefore  - amount;
```

I don't think the alignment is helping. It's a lot of spacing. This just looks better to me:

```suggestion
    assert success => balanceOf(receiver) == balanceBefore + amount;
    assert success => totalSupply() == supplyBefore + amount;
    assert success => underlyingBalanceOf(currentContract) == wrapperUnderlyingBalanceBefore + amount;
    assert success => underlyingBalanceOf(sender) == senderUnderlyingBalanceBefore - amount;
```

---

It honestly looks unreadeable to me

---

feel free to change the spacing/linting